cEP0002.md

```diff
@@ -0,0 +1,272 @@
+during runtime and continues to do so till the task completes.
+7. The main coala process waits on the result queue to get a result and then
+proceeds to process it and log it when it arrives.
+
+# The File Proxy
+
+With the new bear design we are also moving to a new wrapper object to wrap
+every file. This object will also contain some metadata such as the filename
+of the file. The file proxy is an object that contains properties that are
+loaded on demand and cached.
+
+What it will cache:
+    * The filename, this is the starting point for every other attribute.
+    * The content of the file as a string.
+    * The content of the file, line-splitted
+    * It can also cache other things about files as and when required.
+
+The file proxy tries to load from file using the filename if no other cached
+property is already loaded . If the file was already loaded, use the methods
+in the file proxy object to access the splitted lines of the file or the file
+as a string.
+
+A prototype of the file proxy:
+
+```python
+class FileProxy:
+    def __init__(self, filename):
+        self.filename = filename
+        self.loaded = False
+        self.content = []
+
+    def __iter__(self):
+        self.load()
+
+        return iter(self.content)
+
+    def load(self):
+        if self.loaded:
+            return
+
+        with open(filename) as filehandle:
+            self.content = filehandle.readlines()
+```
+
+# Dependency Tracking
+
+A new class will be responsible for keeping track of the dependencies of different bear and it will work in the following manner:
+    * Each bear will add it's dependencies to this tracker class.
+    * Whenever a bear completes it's execution it will send a signal of
+    completion.
+    * This signal would execute the marking of this bear as complete in the
+    dependency tracker.
+    * Once the marking is complete the tracker returns all bears which have no
+    dependencies remaining as pending, so that they can offload tasks to the
+    coala pool
+
+Here is a prototype:
+
+```python
+from sets import Set
+
+class Dependencies:
+    def __init__():
+        self.dependency_dict = {}
+        self.dependency_set = Set()
+
+    def add_dependency(bear_instance, dependency):
+        if bear_instance in self.dependency_dict:
+            self.dependency_dict[bear_instance].append(dependency)
+            self.dependency_set.add(dependency)
+        else:
+            self.dependency_dict[bear_instance] = [dependency]
+            self.dependency_set.add(dependency)
+
+    def resolve(bear_instance):
+        return_bears = []
+        if bear_instance in dependency_set:
+            for bear in self.dependency_dict:
+                self.dependency_dict[bear].remove(bear_instance)
+                if not dependency_dict[bear]:
+                    return_bears.append(bear)
+                    del dependency_dict[bear]
+        return return_bears
```

Comment:
we already have a Dependency class in coala, did you just copy it here? It's nothing new is it?